core/common/src/main/java/alluxio/resource/SharedResourcePool.java

```diff
@@ -0,0 +1,255 @@
+   * `maxAcceptableRefCount`, then we reuse the resource.
+   * 2. If the current size of the resource pool is less than `maxCapacity`, then we
+   * try to create a new resource and use the new one.
+   * 3. If the resource pool is full or creating new one fails, then we fall back to sharing
+   * the least-referenced resource in the pool.
+   * 4. If the resource pool is empty and creating new resource fails, then we throw exception.
+   *
+   * @param maxAcceptableRefCount the maximum degree of acceptable sharing
+   * @return the closeable resource
+   * @throws IOException when the resource pool is empty whereas we failed to create a new one
+   */
+  public CloseableResource<T> acquire(int maxAcceptableRefCount) throws IOException {
+    try (LockResource r = new LockResource(mLock)) {
+      Preconditions.checkState(!mClosed);
+
+      // firstly try to reuse an eligible resource
+      CountingReference<T> ref = mResources.poll();
+      if (ref != null
+          && (ref.getRefCount() <= maxAcceptableRefCount
+              || mResources.size() + 1 >= mMaxCapacity)) {
+        ref.reference();
+        mResources.add(ref);
+        return makeResource(ref);
+      } else if (ref != null) {
+        mResources.add(ref);
+      }
+
+      // secondly try to create a new one
+      try {
+        CountingReference<T> newResource = new CountingReference<>(createNewResource(), 1);
+        mResources.add(newResource);
+        return makeResource(newResource);
+      } catch (IOException e) {
+        if (ref != null) {
+          // create new resource failed, but can fall back to use
+          // shared resource
+          LOG.warn("Reuse shared resource as creating new resource failed", e);
+          mResources.remove(ref);
+          ref.reference();
+          mResources.add(ref);
+          return makeResource(ref);
+        }
+        // can't fall back, error out
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Wraps the inner {@link CountingReference} in a {@link CloseableResource} for return.
+   * @param ref the inner shared reference of the resource
+   * @return the closeable resource for public use
+   */
+  private CloseableResource<T> makeResource(CountingReference<T> ref) {
+    return new CloseableResource<T>(ref.get()) {
+      @Override
+      public void closeResource() {
+        try (LockResource r = new LockResource(mLock)) {
+          mResources.remove(ref);
+          ref.dereference();
+          mResources.add(ref);
+        }
+      }
+    };
+  }
+
+  /**
+   * @return the number of resources the queue currently holds
+   */
+  public int size() {
+    try (LockResource r = new LockResource(mLock)) {
+      Preconditions.checkState(!mClosed);
+
+      return mResources.size();
+    }
+  }
+
+  /**
+   * @return newly created resource
+   * @throws IOException if error occurs
+   */
+  protected abstract T createNewResource() throws IOException;
```

Comment:
why `IOException`?

Reply:
Any suggestions? Is `Exception` more appropriate? `DynamicResourcePool` uses `IOException` so I just copied that.

Reply:
yeah we should use `Exception` here.